Documentation/Tasks/Index.rst

```diff
@@ -0,0 +1,169 @@
+Secondly we see that the task defines a default command if none is specified. If you like you can override this option in your deployment configuration.
+This is valid for all the delivered tasks shipped with Surf.
+
+For this simple task above, we recommend to simplify this by just using the possibility to define your **task dynamically** by the following mechanism provided by Surf::
+
+    <?php
+
+    ...
+
+    $workflow->defineTask('NpmInstallTask', \\TYPO3\\Surf\\Task\\LocalShellTask::class, [
+        'command' => [
+            'cd {workspacePath} && npm install',
+        ]
+    ]);
+
+This way you create a task dynamically by extending the base task \\TYPO3\\Surf\\Task\\LocalShellTask::class with an array of options. In this case the command option is mandatory for the LocalShellTask.
+
+We will show you another convenient and often used way to customize the deployment workflow with your own tasks::
+
+   <?php
+
+   ...
+
+   $workflow->defineTask('CopyEnvFileTask', \\TYPO3\\Surf\\Task\\ShellTask::class, [
+        'command' => [
+            "cp {sharedPath}/.env {releasePath}/.env",
+            "cd {releasePath}",
+        ]
+    ]);
+
+In this case we create a task dynamically based on the ShellTask. The ShellTask execute one or more provided commands on the target machine.
+As you can see, we have used some other placeholders compared to the LocalShellTask above.
+For the remote ShellTask the following placeholders are available:
+
+* **workspacePath**: The path to the local workspace directory
+* **deploymentPath**: The path to the deployment base directory
+* **releasePath**: The path to the release directory in work (typically referenced by *next*)
+* **sharedPath**: The path to the shared directory for all releases
+* **currentPath**: The path that points to the *current* release
+* **previousPath**: The path that points to the *previous* release
+
+Add task to the deployment flow
+-------------------------------
+
+So we have seen how to create custom tasks in different ways. In the following we will see how we add these tasks to the deployment flow::
+
+   <?php
+
+   ...
+
+   $application = new \\TYPO3\\Surf\\Application\\TYPO3\\CMS();
+   $deployment->addApplication($application);
+   $workflow = $deployment->getWorkflow();
+
+   $workflow->defineTask('CopyEnvFileTask', \\TYPO3\\Surf\\Task\\ShellTask::class, [
+      'command' => [
+         "cp {sharedPath}/.env {releasePath}/.env",
+         "cd {releasePath}",
+      ]
+   ]);
+
+   $deployment->onInitialize(function() use ($workflow, $application) {
+      $workflow->afterStage('transfer', 'CopyEnvFileTask', $application);
+   });
+
+This will execute the new task after the stage transfer only for the application referenced by $application.
+
+Besides specifying the execution point via a stage, you can also give an existing task as an anchor and specify the task execution with **afterTask** or **beforeTask**::
+
+   <?php
+
+   ....
```

Comment:
should we add a hint for require statements here?

cases: 
1. phar archive, require in the beginning of the Deployment Definition
2. composer installation, autoloader
3. others?